### PR TITLE
Fix FeedExporter not to export empty file

### LINKED ISSUE #872: FileFeedStorage creates empty file when no items are scraped
When no items are scraped, the corresponding file is created none the less, because it is created in by the `storage.open` call in `FeedExporter.open_spider`. This behavior ignores the the setting of `FEED_STORE_EMPTY` when using file export.

My proposal for this would be to add a `cleanup` method to the `IFeedStorage` interface. Then  `FeedExporter.close_spider` can call that method before returning in case `slot.itemcount` is zero and `self.store_empty` is `False`. `cleanup` could also be called internally from the `store` methods of the `IFeedStorage` implementations.

### LINKED ISSUE #5633: Address BatchDeliveriesTest.test_batch_path_differ failures on Windows
After https://github.com/scrapy/scrapy/pull/5632, the only remaining test failure seems to be that of `BatchDeliveriesTest.test_batch_path_differ`, which seems to be failing in Windows for Python 3.7 (and pinned dependencies), Python 3.9+, but not Python 3.8, for some reason.

It is also weird that the error seems to suggest an extra item is yielded (4 != 3) in those scenarios. I cannot imagine what could cause an extra item yield in a specific platform.

### PR DESCRIPTION
**Pull request by namelessGonbai**
Fixes #872, fixes #5633 
+ Fix FeedExporter not to export empty file
+ Change default value of FEED_STORE_EMPTY

There are several problems with this patch.

1. There is no additional test.

The explanation for the FeedExporter fix is within the scope of what is already done in the existing tests, so we believe that the fixes to the existing tests are sufficient for testing.
However, I have no idea how to test for changes in defaults, even though I think the magnitude of the effect is clearly significant and needs to be tested.

2. Meaning of `test_export_no_items_multiple_feeds()`

This test from #4621.
This test seems to be testing whether a file descriptor leak has been resolved when `FEED_STORE_EMPTY=False`.
`FEED_STORE_EMPTY=False` will no longer open the file if the item is missing and this test is no longer meaningful.
This test needs to be fixed (or deleted) because I made a fix that twisted its original meaning to pass the test.

### CODE REVIEW

**Comment by namelessGonbai** [comment]
I forgot to mention this as well. Another issue with this patch is whether changing the default values is acceptable.

**Comment by namelessGonbai** [comment]
I forgot to state this explicitly.
I would like a review of the problems I listed above.

File: scrapy/settings/default_settings.py [hunk 1]
```diff
@@ -141,7 +141,7 @@
 FEED_TEMPDIR = None
 FEEDS = {}
 FEED_URI_PARAMS = None  # a function to extend uri arguments
-FEED_STORE_EMPTY = False
+FEED_STORE_EMPTY = True
```
Comment (Gallaecio) [inline scrapy/settings/default_settings.py:144]: We need to update [the docs](https://github.com/scrapy/scrapy/blob/d60b4edd11436e61284615ec7ce89f8ac7e46d9a/docs/topics/feed-exports.rst?plain=1#L555) accordingly.

**Review by Gallaecio — COMMENTED** [review]
> I have no idea how to test for changes in defaults, even though I think the magnitude of the effect is clearly significant and needs to be tested.

I don’t think we need a specific test for defaults, I think your test changes already cover the default value change as far as we need.

> This test needs to be fixed (or deleted) because I made a fix that twisted its original meaning to pass the test.

I agree, and I think changing expectations is the way to go:

```diff
--- a/tests/test_feedexport.py
+++ b/tests/test_feedexport.py
@@ -1057,13 +1057,13 @@ class FeedExportTest(FeedExportTestBase):
                 self._random_temp_filename(): {"format": "csv"},
             },
             "FEED_STORAGES": {"file": LogOnStoreFileStorage},
+            "FEED_STORE_EMPTY": False,
         }
 
         with LogCapture() as log:
             yield self.exported_no_data(settings)
 
-        print(log)
-        self.assertEqual(str(log).count("Storage.store is called"), 3)
+        self.assertEqual(str(log).count("Storage.store is called"), 0)
 
     @defer.inlineCallbacks
     def test_export_multiple_item_classes(self):
```

File: scrapy/settings/default_settings.py (hunk 1 above)
Comment (namelessGonbai) [reply scrapy/settings/default_settings.py:144]: I forgot about that.
Other than changing the default values, do I need to make any additions to the documentation with this modification? For example, mentioning that files are not exported explicitly, etc.

Comment (Gallaecio) [reply scrapy/settings/default_settings.py:144]: To mention that files are not exported may be a good thing, yes.

I don’t think we need anything else, though. We will need to justify this default value change, explain the reasons behind it, in the release notes, but we usually handle release notes entries when preparing the release, so you do not need to worry about that part here.

Comment (namelessGonbai) [reply scrapy/settings/default_settings.py:144]: Understood. Thank you very much.

File: docs/topics/feed-exports.rst [hunk 2]
```diff
@@ -552,9 +552,10 @@ to ``.json`` or ``.xml``.
 FEED_STORE_EMPTY
 ----------------
 
-Default: ``False``
+Default: ``True``
 
 Whether to export empty feeds (i.e. feeds with no items).
+If False and there is no items, no new files are created and existing files are not modified.
```
Comment (Gallaecio) [inline docs/topics/feed-exports.rst:558]:
```suggestion
If ``False``, and there are no items to export, no new files are created and 
existing files are not modified, even if the :ref:`overwrite feed option 
<feed-options>` is enabled.
```

**Comment by namelessGonbai** [comment]
Findings.
The test had a potential bug.
https://github.com/scrapy/scrapy/blob/6ab49e954f25d491df7986065d270bb0068c7c89/tests/test_feedexport.py#L2500-L2508
This will overwrite the same file when the file write is fast enough and the batch_time matches.

**Comment by namelessGonbai** [comment]
I can rewrite the whole test to %(batch_id)d if you need to fix it at least, and I'm sure it's the type of error that disappears when you re-run the runner in the first place, though.

**Comment by Gallaecio** [comment]
I think I have seen that test fail before, I would not worry about it here, fixing it is out of the scope of this change.

**Review by Gallaecio — APPROVED** [review]

**Comment by namelessGonbai** [comment]
No merge or additional review yet, any problems?

File: scrapy/extensions/feedexport.py [hunk 3]
```diff
@@ -277,18 +277,21 @@ def _store_in_thread(self, file):
 class FeedSlot:
     def __init__(
         self,
-        file,
-        exporter,
```
Comment (wRAR) [inline scrapy/extensions/feedexport.py:281]: This now changes a public class, as we made FeedSlot public recently. @kmike @Gallaecio thoughts?

Comment (Gallaecio) [reply scrapy/extensions/feedexport.py:281]: The only backward-compatible option I can think of at the moment is to keep the current class unchanged, deprecate it, and create a new class with the new API with a new class name (`BatchSlot`?).

Comment (wRAR) [reply scrapy/extensions/feedexport.py:281]: We decided that as this only changes the `__init__` signature and the current usage of this class is to receive it in the `feed_slot_closed` signal handler, it's fine to just change this.

**Review by wRAR — APPROVED** [review]

**Review by Gallaecio — APPROVED** [review]

**Comment by kmike** [comment]
Thanks @namelessGonbai, @wRAR and @Gallaecio!

**Pull request merged**